src/auth/src/credentials/mds_credential.rs

```
@@ -58,7 +126,29 @@ where
         let mut value = HeaderValue::from_str(&format!("{} {}", token.token_type, token.token))
             .map_err(CredentialError::non_retryable)?;
         value.set_sensitive(true);
-        Ok(vec![(AUTHORIZATION, value)])
+        let mut headers = vec![(AUTHORIZATION, value)];
+        if let Some(project) = &self.quota_project_id {
+            headers.push((
+                HeaderName::from_static(QUOTA_PROJECT_KEY),
+                HeaderValue::from_str(project).map_err(CredentialError::non_retryable)?,
+            ));
+        }
+        Ok(headers)
+    }
+
+    async fn get_universe_domain(&self) -> Option<String> {
+        {
+            let universe_domain = self.universe_domain.read().unwrap();
+            if universe_domain.is_some() {
+                return universe_domain.clone();
+            }
+        }
+        let universe_domain = match self.get_universe_domain_from_mds().await {
```

It is possible that multiple threads kick off this mds request. You would have to put it behind a lock so that only one thread will call MDS.

---

Seems to me like this is the same problem we have in the token cache.  Should we solve this once in one place? Or at least use the same solution?

---

We need to keep refreshing tokens but for UD, it is one time thing. So we do not need the forever background task.

For this, maybe we could do a simplified form of that: 
1. Create a watch channel for "storing" universe domain: Give transmitter to background task and receiver is used by get_universe_domain.
2. Create another watch channel (maybe oneshot will work, need to investigate) for informing the background task. Lets calls this wakeup_channel.
3. Background task will wait to be woken up by a message in the wakeup_channel. Once woken up, it obtains the UD from MDS, pushes it onto the ud_channel and quits.
4. `get_universe_domain` will first check receiver of ud_channel. If a value was available in it, that is returned. If it did not find a value there, it will push a message to wakeup_channel and then wait on a changed notification on the ud_channel. Once the background task provides the ud value on the channel, it wakes up, reads and returns it.

---

thanks for the detailed explanation, I think oneshot will work here, just if the call happens concurrently then we would ignore error and thread would go onto waiting for the result to come back from the endpoint.